e2e/op_stack.go

```diff
@@ -0,0 +1,282 @@
+	// LogWithPrefix is meant to be used as a general-purpose log function.
+	// The prefix will never be empty and r will never be nil.
+	// It is only used to fulfill log interfaces defined by third-party OP Stack components.
+	LogWithPrefix(prefix string, r *log.Record)
+	AfterStartup()
+	BeforeShutdown()
+}
+
+type OPStack struct {
+	// The http address (without the http://) of the Ethereum EL node.
+	l1Address string
+	// The websocket address (without the ws://) of the execution engine.
+	engineAddress string
+	// The http address (without the http://) of the op-node.
+	nodeAddress         string
+	privKey             *ecdsa.PrivateKey
+	rollupConfig        *rollup.Config
+	l2OutputOracleProxy common.Address
+	eventListener       OPEventListener
+}
+
+// TODO setup verifiers
+
+func NewOPStack(
+	l1Address,
+	engineAddress,
+	opNodeAddress string,
+	l2OutputOracleProxy common.Address,
+	privKey *ecdsa.PrivateKey,
+	rollupConfig *rollup.Config,
+	eventListener OPEventListener,
+) *OPStack {
+	return &OPStack{
+		l1Address:           l1Address,
+		engineAddress:       engineAddress,
+		nodeAddress:         opNodeAddress,
+		privKey:             privKey,
+		rollupConfig:        rollupConfig,
+		l2OutputOracleProxy: l2OutputOracleProxy,
+		eventListener:       eventListener,
+	}
+}
+
+func (op *OPStack) Run(parentCtx context.Context) (err error) {
+	var wg conc.WaitGroup
+	defer wg.Wait()
+	ctx, cancel := context.WithCancelCause(parentCtx)
+	defer func() {
+		op.eventListener.BeforeShutdown()
+		cancel(err)
+		err = utils.Cause(ctx)
+	}()
+
+	anvilRPCClient, err := rpc.DialContext(ctx, "http://"+op.l1Address)
+	if err != nil {
+		return fmt.Errorf("dial anvil: %v", err)
+	}
+	anvil := NewAnvil(anvilRPCClient)
+
+	// Run op-node.
+	wg.Go(func() {
+		cancel(op.runNode(ctx))
+	})
+
+	// Use the same tx manager config for the op-proposer and op-batcher.
+	defaults := txmgr.DefaultBatcherFlagValues
+	l1ChainID, err := anvil.ChainID(ctx)
+	if err != nil {
+		return fmt.Errorf("get l1 chain id: %v", err)
+	}
+	txManagerConfig := &txmgr.Config{
+		Backend:                   anvil,
+		ChainID:                   l1ChainID,
+		NumConfirmations:          defaults.NumConfirmations,
+		NetworkTimeout:            defaults.NetworkTimeout,
+		FeeLimitMultiplier:        defaults.FeeLimitMultiplier,
+		ResubmissionTimeout:       defaults.ResubmissionTimeout,
+		ReceiptQueryInterval:      defaults.ReceiptQueryInterval,
+		TxNotInMempoolTimeout:     defaults.TxNotInMempoolTimeout,
+		SafeAbortNonceTooLowCount: defaults.SafeAbortNonceTooLowCount,
+		Signer: func(ctx context.Context, address common.Address, tx *ethtypes.Transaction) (*ethtypes.Transaction, error) {
+			return opcrypto.PrivateKeySignerFn(op.privKey, l1ChainID)(address, tx)
+		},
+	}
+
+	wg.Go(func() {
+		cancel(op.runProposer(ctx, anvil, txManagerConfig))
+	})
+
+	wg.Go(func() {
+		cancel(op.runBatcher(ctx, anvil, txManagerConfig))
+	})
+
+	op.eventListener.AfterStartup()
+	<-ctx.Done()
+	return nil
```

Comment:
Should these be `return err` for most of these functions? They all declare named returns, and some of them set the `err` variable during a deferred cleanup.

Reply:
`err` will be set to the value that the function returns with. I find `return nil` to be more explicit than `return err`, since the latter makes one think that `err != nil` somehow. Your thoughts?

Reply:
I wrote that believing that the below would return `nil`. It doesn't, which honestly I find a little disturbing, but I'm adjusting.

I think it can stay as is, or possibly the return statement could be dropped altogether.

```
func impErr() (err error) {
	defer func() {
		err = errors.New("this is an error")
	}()

	return nil
}
```

Reply:
I see. The rabbit ai bot thought it was confusing as well.

I find it tough to return coherent error messages when dealing with concurrency in general. Using named returns, defers, and a `context.CancelCauseFunc` is the best solution I could think of.